### PR TITLE
Add method to wait until an update to a table has been applied before continuing

### PR DESCRIPTION
**Pull request by dylanclementas**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -658,6 +658,30 @@ DynamoTable.prototype.createTableAndWait = function(readCapacity, writeCapacity,
...
+  options.TableName = options.TableName || this.name
+  var self = this
+
+  this.describeTable(function(err, data) {
+    if (err) return cb(err)
+    if (data.TableStatus !== 'ACTIVE') return setTimeout(self.updateTableAndWait(readCapacity, writeCapacity, options, cb), 1000)
+
+    // If the throughput has been provisoned then return
+    if (readCapacity && data.ProvisionedThroughput.ReadCapacityUnits == readCapacity
+      && writeCapacity && data.ProvisionedThroughput.WriteCapacityUnits == writeCapacity) {
```
Comment (mhart) [inline index.js]: Move the first `&&` to the end of the last line

**Comment by mhart** [comment]
Awesome stuff! Not entirely sure the best strategy is to call `describeTable` first though - my first thought would be to call update first, then call `describeTable` until it's updated... thoughts?

**Comment by dylanclementas** [comment]
Had a look @mhart, the reason the `describeTable` call is done first is because we are calling the method recursively when `setTimeout` fires. 

We can change it around but that would probably mean creating a new method to call on time out. Doing it this way is a bit cleaner imho, let me know if you disagree and I can change it.

**Comment by mhart** [comment]
Yeah I disagree :-) Just having the describe first because it's a little shorter to write the recursion still doesn't seem like it passes the self-describing-function test for me.

**Comment by dylanclementas** [comment]
@mhart I have updated the code as per your suggestion.

File: index.js [hunk 2]
```diff
@@ -658,6 +658,37 @@ DynamoTable.prototype.createTableAndWait = function(readCapacity, writeCapacity,
...
+
+  self.updateTable(readCapacity, writeCapacity, options, function(err) {
+    if (err) return cb(err)
+
+    // check whether the update has in fact been applied
+    function checkTable(count) {
+      // Make sure we don't go into an infinite loop
+      if (++count > 1000) return cb(new Error('Wait limit exceeded'))
+
+      var self = this
```
Comment (mhart) [inline index.js]: No need for this

File: index.js [hunk 3]
```diff
@@ -658,6 +658,37 @@ DynamoTable.prototype.createTableAndWait = function(readCapacity, writeCapacity,
...
+        if (data.TableStatus !== 'ACTIVE') return setTimeout(checkTable.bind(self, count), 1000)
+
+        // If the throughput has been provisoned then return
+        if (readCapacity && data.ProvisionedThroughput.ReadCapacityUnits === readCapacity &&
+          writeCapacity && data.ProvisionedThroughput.WriteCapacityUnits === writeCapacity) {
+          return cb(null, data)
+        }
+      })
+    }
+    checkTable.bind(self)(1)
```
Comment (mhart) [inline index.js]: Or the `bind` here

File: index.js [hunk 4]
```diff
@@ -658,6 +658,37 @@ DynamoTable.prototype.createTableAndWait = function(readCapacity, writeCapacity,
...
+
+    // check whether the update has in fact been applied
+    function checkTable(count) {
+      // Make sure we don't go into an infinite loop
+      if (++count > 1000) return cb(new Error('Wait limit exceeded'))
+
+      var self = this
+      self.describeTable(function(err, data) {
+        if (err) return cb(err)
+        if (data.TableStatus !== 'ACTIVE') return setTimeout(checkTable.bind(self, count), 1000)
```
Comment (mhart) [inline index.js]: No need to bind to `self` here either (can just be `null`)

Comment (mhart) [inline index.js]: Actually, can just do `setTimeout(checkTable, 1000, count)`

File: index.js [hunk 5]
```diff
@@ -658,6 +658,37 @@ DynamoTable.prototype.createTableAndWait = function(readCapacity, writeCapacity,
...
+
+        // If the throughput has been provisoned then return
+        if (readCapacity && data.ProvisionedThroughput.ReadCapacityUnits === readCapacity &&
+          writeCapacity && data.ProvisionedThroughput.WriteCapacityUnits === writeCapacity) {
```
Comment (mhart) [inline index.js]: Indent this one more block

**Comment by mhart** [comment]
Cool! Just a couple of unnecessary binds in there - also feel free to squash

File: index.js (hunk 4 above)
Comment (dylanclementas) [inline index.js]: nice :+1:

**Comment by dylanclementas** [comment]
@mhart done. Do I need to bump the version as well?

**Comment by mhart** [comment]
@dylanclementas nah I'll do that

**Comment by mhart** [comment]
Very nice! Merging :+1:

**Comment by mhart** [comment]
Published as v0.8.0

**Pull request merged**